File: Hash_table/python/hash_table.py
import numpy as np

class Node:
    def __init__(self, id, name):
        self.key = id  # Store the key (ID)
        self.val = name  # Store the value (name)
        self.next = None  # Pointer to next node for handling collisions

class HashTable:
    def __init__(self, size):
        self.size = size  # Size of the hash table
        self.tableArr = np.empty(size, dtype=object)  # Initialize array with None values
    
    def hashFunction(self, key):
        return key % self.size  # Simple modulo-based hash function
    
    def insert(self, id, name):
        hashIndx = self.hashFunction(id)
        newNode = Node(id, name)

        # If no collision, insert directly
        if self.tableArr[hashIndx] is None:
            self.tableArr[hashIndx] = newNode
        else:
            # Handle collision using chaining (linked list)
            temp = self.tableArr[hashIndx]
            while temp.next is not None:
                temp = temp.next
            temp.next = newNode
    
    def printTable(self):
        print('='*55)
        for i in range(len(self.tableArr)):
            print(i, end=' : ')
            temp = self.tableArr[i]
            while temp is not None:
                print(f'({temp.key}, {temp.val}) -> ', end='')
                temp = temp.next
            print('None')
        print('='*55)
    
    def search(self, id):
        hashIndx = self.hashFunction(id)
        temp = self.tableArr[hashIndx]
        
        while temp is not None:
            if temp.key == id:
                return hashIndx  # Return index if key is found
            temp = temp.next
        return -1  # Key not found
    
    def remove(self, id):
        hashIndx = self.hashFunction(id)
        temp = self.tableArr[hashIndx]
        
        if temp is  None: # If there's no node available to remove
            return
        
        if temp.key == id: # If the node to be removed is the first node
            self.tableArr[hashIndx] = temp.next
            return
        
        # Traverse the linked list to find and remove the node
        while temp.next is not None:
            if temp.next.key == id:
                temp.next = temp.next.next
                return
            temp = temp.next

# Test cases to validate HashTable functionality
if __name__ == "__main__":
    hashTable1 = HashTable(5)
    
    # Insert elements
    hashTable1.insert(101, 'Bob')
    hashTable1.insert(102, 'Ben')
    hashTable1.insert(111, 'Stewie')
    hashTable1.insert(106, 'Anna')  # Testing collision (106 % 5 == 1, same as 101)
    hashTable1.insert(201, 'Chris') # Another collision case
    
    # Print table before removal
    print("Hash Table before deletion:")
    hashTable1.printTable()
    
    # Search tests
    print("Search 101:", hashTable1.search(101))  # Expected: valid index
    print("Search 103:", hashTable1.search(103))  # Expected: -1 (not found)
    
    # Remove elements
    hashTable1.remove(101) # Remove an existing key
    hashTable1.remove(103) # Attempt to remove a non-existing key
    
    # Print table after removal
    print("Hash Table after deletion:")
    hashTable1.printTable()

File: Hash_table/java/hash_table.java
class Node {
    int key;
    String val;
    Node next;

    // Constructor to initialize a node with key-value pair
    Node(int id, String name) {
        this.key = id;
        this.val = name;
        this.next = null;
    }
}

class HashTable {
    Node[] tableArr; // Array of nodes representing the hash table
    int size; // Size of the hash table

    // Constructor to initialize hash table with a given size
    HashTable(int size) {
        tableArr = new Node[size];
        this.size = size;
    }

    // Hash function to determine index for a given key
    public int hashFunction(int id) {
        return id % this.size;
    }

    // Insert a new key-value pair into the hash table
    public void insert(int id, String name) {
        int hashIndx = hashFunction(id);
        Node newNode = new Node(id, name);

        // If no collision, insert directly
        if (this.tableArr[hashIndx] == null) {
            this.tableArr[hashIndx] = newNode;
        } else {
            // Handle collision using chaining (linked list)
            Node temp = this.tableArr[hashIndx];
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = newNode;
        }
    }

    // Search for a key in the hash table
    public boolean search(int id) {
        int hashIndx = hashFunction(id);
        Node temp = this.tableArr[hashIndx];

        while (temp != null) {
            if (temp.key == id) {
                return true; // Key found
            }
            temp = temp.next;
        }
        return false; // Key not found
    }

    // Remove a key from the hash table
    public void remove(int id) {
        int hashIndx = hashFunction(id);
        Node temp = this.tableArr[hashIndx];

        if (temp == null)
            return; // Key not found

        // If the key to be removed is the first node
        if (temp.key == id) {
            this.tableArr[hashIndx] = temp.next;
            return;
        }

        // Traverse the linked list to find and remove the key
        while (temp.next != null) {
            if (temp.next.key == id) {
                temp.next = temp.next.next;
                return;
            }
            temp = temp.next;
        }
    }

    // Print the hash table
    public void printTable() {
        System.out.println("=".repeat(50));

        for (int i = 0; i < tableArr.length; i++) {
            System.out.print(i + " : ");
            Node temp = tableArr[i];
            while (temp != null) {
                System.out.print("(" + temp.key + " , " + temp.val + ") -> ");
                temp = temp.next;
            }
            System.out.println("null");
        }
        
        System.out.println("=".repeat(50));

    }
}

public class hash_table {
    public static void main(String[] args) {
        HashTable h1 = new HashTable(5);

        // Insert elements
        h1.insert(101, "Bob");
        h1.insert(102, "Steve");
        h1.insert(111, "Paige");
        h1.insert(106, "Anna"); // Testing collision (106 % 5 == 1, same as 101)
        h1.insert(201, "Chris"); // Another collision case

        // Search tests
        System.out.println("Search 101: " + h1.search(101)); // Expected: true
        System.out.println("Search 103: " + h1.search(103)); // Expected: false

        // Print table before removal
        System.out.println("Hash Table before deletion:");
        h1.printTable();

        // Remove elements
        h1.remove(101); // Remove an existing key
        h1.remove(103); // Attempt to remove a non-existing key

        // Print table after removal
        System.out.println("Hash Table after deletion:");
        h1.printTable();
    }
}
